ddl/ddl_api.go

```diff
@@ -2305,6 +2329,129 @@ func (d *ddl) AddColumn(ctx sessionctx.Context, ti ast.Ident, spec *ast.AlterTab
+		for _, specNewColumn := range spec.NewColumns {
+			newColumnsCount++
+			if !set[specNewColumn.Name.Name.O] {
+				set[specNewColumn.Name.Name.O] = true
+			} else {
+				return errors.Trace(infoschema.ErrColumnExists.GenWithStackByArgs(specNewColumn.Name.Name.O))
+			}
+		}
+	}
+	if err = checkAddColumnTooManyColumns(len(t.Cols()) + newColumnsCount); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Check the columns one by one
+	for _, spec := range specs {
+		for _, specNewColumn := range spec.NewColumns {
+			err := checkUnsupportedColumnConstraint(specNewColumn, ti)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			colName := specNewColumn.Name.Name.O
+			if err = checkColumnAttributes(colName, specNewColumn.Tp); err != nil {
+				return errors.Trace(err)
+			}
+			if len(colName) > mysql.MaxColumnNameLength {
+				return ErrTooLongIdent.GenWithStackByArgs(colName)
+			}
+
+			// If new column is a generated column, do validation.
+			// NOTE: we do check whether the column refers other generated
+			// columns occurring later in a table, but we don't handle the col offset.
+			for _, option := range specNewColumn.Options {
+				if option.Tp == ast.ColumnOptionGenerated {
+					if err := checkIllegalFn4GeneratedColumn(specNewColumn.Name.Name.L, option.Expr); err != nil {
+						return errors.Trace(err)
+					}
+
+					if option.Stored {
+						return errUnsupportedOnGeneratedColumn.GenWithStackByArgs("Adding generated stored column through ALTER TABLE")
+					}
+
+					_, dependColNames := findDependedColumnNames(specNewColumn)
+					if err = checkAutoIncrementRef(specNewColumn.Name.Name.L, dependColNames, t.Meta()); err != nil {
+						return errors.Trace(err)
+					}
+					duplicateColNames := make(map[string]struct{}, len(dependColNames))
+					for k := range dependColNames {
+						duplicateColNames[k] = struct{}{}
+					}
+					cols := t.Cols()
+
+					if err = checkDependedColExist(dependColNames, cols); err != nil {
+						return errors.Trace(err)
+					}
+
+					if err = verifyColumnGenerationSingle(duplicateColNames, cols, spec.Position); err != nil {
+						return errors.Trace(err)
+					}
+				}
+			}
+
+			// Check whether added column has existed.
+			col := table.FindCol(t.Cols(), colName)
+			if col != nil {
+				err = infoschema.ErrColumnExists.GenWithStackByArgs(colName)
+				if spec.IfNotExists {
+					ctx.GetSessionVars().StmtCtx.AppendNote(err)
+					return nil
+				}
+				return err
+			}
+
+			// Ignore table constraints now, maybe return error later.
+			// We use length(t.Cols()) as the default offset firstly, we will change the
+			// column's offset later.
+			col, _, err = buildColumnAndConstraint(ctx, len(t.Cols()), specNewColumn, nil, t.Meta().Charset, t.Meta().Collate, schema.Charset, schema.Collate)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			col.OriginDefaultValue, err = generateOriginDefaultValue(col.ToInfo())
+			if err != nil {
+				return errors.Trace(err)
+			}
```

Comment:
It seems that the logic of line2362-2430 is similar to some logic in the `AddColumn` function. Could we extract a function to handle the duplicated codes?

Reply:
Done